hbase_storage/src/main/java/com/splicemachine/si/impl/RollBackTxnFilter.java

```diff
@@ -0,0 +1,102 @@
+import com.splicemachine.si.api.txn.TxnSupplier;
+import com.splicemachine.si.api.txn.TxnView;
+import com.splicemachine.si.constants.SIConstants;
+import com.splicemachine.si.impl.driver.SIDriver;
+import com.splicemachine.storage.*;
+import com.splicemachine.utils.ByteSlice;
+import com.splicemachine.utils.SpliceLogUtils;
+import org.apache.hadoop.hbase.Cell;
+import org.apache.hadoop.hbase.filter.Filter;
+import org.apache.hadoop.hbase.filter.FilterBase;
+import org.apache.hadoop.hbase.util.Bytes;
+import org.apache.log4j.Logger;
+
+import java.util.List;
+
+public class RollBackTxnFilter extends FilterBase {
+
+    private static final Logger LOG=Logger.getLogger(RollBackTxnFilter.class);
+
+
+    private List<DataMutation> mutations;
+    private OperationFactory opFactory;
+    private TxnSupplier txnSupplier;
+
+    public RollBackTxnFilter(TxnSupplier txnSupplier, List<DataMutation> mutations) {
+        this.mutations = mutations;
+        opFactory = SIDriver.driver().baseOperationFactory();
+        this.txnSupplier = txnSupplier;
+    }
+
+    @Override
+    public Filter.ReturnCode filterCell(Cell keyValue){
+        DataCell data = new HCell(keyValue);
+        long txnId = data.version();
+        try {
+
+            if (txnSupplier != null) {
+                TxnView txn = txnSupplier.getTransaction(txnId);
+                if (txn.getEffectiveState() == Txn.State.ROLLEDBACK) {
+                    if (LOG.isDebugEnabled()) {
+                        CellType type = data.dataType();
+                        byte[] k = data.key();
+                        byte[] v = data.value();
+                        SpliceLogUtils.debug(LOG, "Filter out %s, k=%s, v=%s", type.toString(),
+                                Bytes.toStringBinary(k), Bytes.toStringBinary(v));
+
+                    }
+                    ByteSlice bs = new ByteSlice();
+                    bs.set(keyValue.getRowArray(), keyValue.getRowOffset(), keyValue.getRowLength());
+                    DataDelete delete = opFactory.newDelete(bs);
+                    delete.deleteColumn(SIConstants.DEFAULT_FAMILY_BYTES,SIConstants.FK_COUNTER_COLUMN_BYTES, keyValue.getTimestamp());
+                    delete.deleteColumn(SIConstants.DEFAULT_FAMILY_BYTES,SIConstants.FIRST_OCCURRENCE_TOKEN_COLUMN_BYTES, keyValue.getTimestamp());
+                    delete.deleteColumn(SIConstants.DEFAULT_FAMILY_BYTES,SIConstants.PACKED_COLUMN_BYTES, keyValue.getTimestamp());
+                    delete.deleteColumn(SIConstants.DEFAULT_FAMILY_BYTES,SIConstants.TOMBSTONE_COLUMN_BYTES, keyValue.getTimestamp());
+                    delete.deleteColumn(SIConstants.DEFAULT_FAMILY_BYTES,SIConstants.COMMIT_TIMESTAMP_COLUMN_BYTES, keyValue.getTimestamp());
+                    delete.addAttribute(SIConstants.SUPPRESS_INDEXING_ATTRIBUTE_NAME,SIConstants.SUPPRESS_INDEXING_ATTRIBUTE_VALUE);
+                    delete.setDurability(Durability.NONE);
+                    mutations.add(delete);
```

Comment:
this duplicates the logic in `com.splicemachine.si.impl.server.ConflictRollForward#handle`, instead:
- pass in the `ConflictRollForward` object to this class constructor (instead of the `mutation`)
- let it generate the deletion mutation itself by calling `conflictRowForward.handle(data)`.

IMHO this makes the intention clearer and reduces code duplication.

Reply:
It's not exactly same as ConflictRollForward#handle